### PR TITLE
Add ACPI support for x86_64

### LINKED ISSUE #2493: [ACPI] Add support for ACPI Power Management
# Feature Request

This is a feature tracking issue for the work to enable power management support for Firecracker microVMs through ACPI.

## Describe the desired solution

Implement virtual power management via ACPI which will result in the usual power related commands (reboot, halt, shutdown) not needing a dedicated device (i.e i8042 on x86_64).

## Describe possible alternatives

No other alternatives are taken into account.

## Checks

- [ ] Have you searched the Firecracker Issues database for similar requests?
- [ ] Have you read all the existing relevant Firecracker documentation?
- [ ] Have you read and understood Firecracker's core tenets?

### PR DESCRIPTION
**Pull request by bchalios**
## Changes

Introduce ACPI support for x86_64.

## Reason

Introducing ACPI will let us do a couple of things:
* Get rid of the i8042 (minimal) emulation. We will be able to shutdown by means of the S5 state. When/If we integrate ACPI for ARM, we will be able to shutdown properly on ARM as well
* Allow declaring devices without injecting slugs in kernel's command line.
* ACPI is necessary if we want to introduce support for PCIe
* Possible enabler for device/cpu/mem hotplugging
* Possible enabler for expressing NUMA topologies

Closes #2493

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] New `unsafe` code is documented.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md

### CODE REVIEW

File: src/vmm/src/device_manager/mmio.rs [hunk 1]
```diff
@@ -102,39 +103,30 @@ pub struct MMIODeviceInfo {
...
             id_to_dev_info: HashMap::new(),
         })
     }
 
     /// Allocates resources for a new device to be added.
-    fn allocate_mmio_resources(&mut self, irq_count: u32) -> Result<MMIODeviceInfo> {
-        let irqs = (0..irq_count)
-            .map(|_| self.irq_allocator.allocate_id())
-            .collect::<vm_allocator::Result<_>>()
+    pub(crate) fn allocate_mmio_resources(
```
Comment (zulinx86) [inline src/vmm/src/device_manager/mmio.rs:119]:
```suggestion
    fn allocate_mmio_resources(
```
As per review on #3124 , we don't need to make this public.
This associated function is used only inside this file.
Keeping this private can avoid misuse (e.g., only allocates MMIO resources, but not register them to KVM).

**Review by zulinx86 — CHANGES_REQUESTED** [review]

File: src/vmm/src/device_manager/mmio.rs (hunk 1 above)
Comment (bchalios) [reply src/vmm/src/device_manager/mmio.rs]: You are right! Ty!

**Comment by bchalios** [comment]
Mostly a note to self, we need to update ci artifacts with kernels that include ACPI support. Otherwise we'll keep seeing the CI failing.

File: src/acpi/src/arch/x86_64.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+use vm_memory::ByteValued;
+
+use crate::fadt::Fadt;
+
+// Disable VGA probing
+const IAPC_BOOT_ARG_FLAGS_VGA_NOT_PRESENT: u8 = 2;
+// Do not enable MSI
+const IAPC_BOOT_ARG_FLAGS_MSI_NOT_PRESENT: u8 = 3;
+// Do not enable ASPM control
+const IAPC_BOOT_ARG_FLAGS_PCI_ASPM: u8 = 4;
```
Comment (JonathanWoollett-Light) [inline src/acpi/src/arch/x86_64.rs:15]: Would it not be easier to define them as their masks?
```suggestion
// Disable VGA probing
const IAPC_BOOT_ARG_FLAGS_VGA_NOT_PRESENT: u8 = 1 << 2;
// Do not enable MSI
const IAPC_BOOT_ARG_FLAGS_MSI_NOT_PRESENT: u8 = 1 << 3;
// Do not enable ASPM control
const IAPC_BOOT_ARG_FLAGS_PCI_ASPM: u8 = 1 << 4;
```

File: src/acpi/src/arch/x86_64.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        Vec::with_capacity(num_cpus * size_of::<LocalAPIC>() + size_of::<IoAPIC>());
+
+    (0..num_cpus).for_each(|cpu_id| {
+        interrupt_controllers.extend(LocalAPIC::new(cpu_id as u8).as_slice());
+    });
+
+    interrupt_controllers.extend(IoAPIC::new(0).as_slice());
+
+    interrupt_controllers
+}
```
Comment (JonathanWoollett-Light) [inline src/acpi/src/arch/x86_64.rs:101]:
```suggestion
pub(crate) fn create_apic_structures(num_cpus: usize) -> Vec<u8> {
    let interrupt_controllers = (0..num_cpus)
        .flat_map(|cpu_id| LocalAPIC::new(cpu_id as u8).as_slice())
        .extend(IoAPIC::new(0).as_slice())
        .collect();

    interrupt_controllers
}
```

**Comment by JonathanWoollett-Light** [comment]
Closes #2493

**Comment by bchalios** [comment]
Closing this in favour of #4428. This other one is based on this one, but rebased on top of our current `main`

**Pull request closed without merging**